pyvrp/cpp/educate/LocalSearch.cpp

```diff
@@ -131,6 +122,30 @@ Individual LocalSearch::intensify(Individual &individual,
     return exportIndividual();
 }
 
+bool LocalSearch::applyNodeOpsWithEmptyRoutes(
+    Node *U, CostEvaluator const &costEvaluator)
+{
+    // Loop over all empty routes that are not exchangable with the previous
+    // empty route (assuming routes are grouped) and try the operator
+    int prev_r = -1;
+    bool success = false;
+    for (size_t r = 0; r < routes.size(); r++)
+    {
+        if (routes[r].empty()
+            && (prev_r == -1 || (data.route(r) != data.route(prev_r))))
+        {
+            prev_r = r;
+            // Note: if the operation is succesful, we still continue
+            // checking operations with other empty routes, similar to
+            // how a move involving U, V will still check U, V' afterwards
+            if (applyNodeOps(U, routes[r].depot, costEvaluator))
+                success = true;
+            // TODO break for loop if data.isHomogeneousFleet()?
+        }
+    }
+    return success;
```

Comment:
I get that these optimisations are possible, but I think we are moving too fast. I would prefer a streamlined implementation like the following, in `LocalSearch::search`:
```cpp
if (step > 0)
  for (auto &route : routes)  // test U and empty routes
    if (route.empty() && applyNodeOps(U, route.depot, costEvaluator))
      break;
```
This is marginally less efficient, but I'd like to see profiling results first before deciding optimisations are needed here.

Reply:
Your code doesn't do the same (it only applies the first improvement then breaks) so it won't get that simple.

Also, I don't have time to go and benchmark every single function which is trivial to optimize, also I don't see how this is unnecessarily complex. 

Feel free to change this if you're convinced the simpler version is not (never!) slower, but I think checking operations for every empty route is potentially extremely wasteful, e.g. in EURO-NeurIPS use case where the number of routes/vehicles was 'unlimited', so equal to the number of customers.